// dedalus-labs-mcp/tools/health/check-health.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.tool = exports.metadata = void 0;
const filtering_1 = require("dedalus-labs-mcp/filtering");
const types_1 = require("dedalus-labs-mcp/tools/types");
exports.metadata = {
    resource: 'health',
    operation: 'read',
    tags: [],
    httpMethod: 'get',
    httpPath: '/health',
    operationId: 'health_check_health_get',
};
exports.tool = {
    name: 'check_health',
    description: "When using this tool, always use the `jq_filter` parameter to reduce the response size and improve performance.\n\nOnly omit if you're sure you don't need the data.\n\nSimple health check.\n\n# Response Schema\n```json\n{\n  type: 'object',\n  title: 'HealthResponse',\n  description: 'Health check response model.',\n  properties: {\n    status: {\n      type: 'string',\n      title: 'Status'\n    }\n  },\n  required: [    'status'\n  ]\n}\n```",
    inputSchema: {
        type: 'object',
        properties: {
            jq_filter: {
                type: 'string',
                title: 'jq Filter',
                description: 'A jq filter to apply to the response to include certain fields. Consult the output schema in the tool description to see the fields that are available.\n\nFor example: to include only the `name` field in every object of a results array, you can provide ".results[].name".\n\nFor more information, see the [jq documentation](https://jqlang.org/manual/).',
            },
        },
        required: [],
    },
    annotations: {
        readOnlyHint: true,
    },
};
const handler = async (client, args) => {
    const { jq_filter } = args;
    return (0, types_1.asTextContentResult)(await (0, filtering_1.maybeFilter)(jq_filter, await client.health.check()));
};
exports.handler = handler;
exports.default = { metadata: exports.metadata, tool: exports.tool, handler: exports.handler };
